xwiki-platform-core/xwiki-platform-oldcore/src/main/java/com/xpn/xwiki/internal/store/hibernate/HibernateStore.java

```diff
@@ -506,6 +422,38 @@ public DatabaseProduct getDatabaseProductName()
         return this.databaseProductCache;
     }
 
+    /**
+     * @return the Hibernate adapter
+     * @throws HibernateException when failing to get the adapter
+     */
+    public HibernateAdapter getAdapter() throws HibernateException
+    {
+        if (this.adapter == null) {
+            this.adapter = createHibernateAdapter();
+        }
+
+        return this.adapter;
+    }
+
+    private synchronized HibernateAdapter createHibernateAdapter()
+    {
+        // Gather metadata about the database
+        DatabaseMetaData metadata = getDatabaseMetaData();
+
+        // Ask the various factories for adapters matching the metadata and configuration
+        for (HibernateAdapterFactory factory : this.adapterFactories) {
+            Optional<HibernateAdapter> newAdapter = factory.createHibernateAdapter(metadata, getConfiguration());
+
+            // If a matching adapter was found, stop searching
+            if (newAdapter.isPresent()) {
+                return newAdapter.get();
+            }
+        }
+
+        // Fallback on the default adapter if not specific one could be found
+        return this.defaultHibernateAdapterProvider.get();
```

Comment:
Sounds like this should be the behaviour of the factory itself.

Reply:
The factories are currently supposed to provide only specific adapters. The current default factory (what you call `the factory` I assume) is a temporary thing with some common behavior, but eventually there will most probably be more (you need one as soon as you need to know more than the database id and version to choose which adapter to return, I even hesitated to have two PostgreSQL adapters for the database and schema modes, and we probably will eventually).